src/snowflake/snowpark/dataframe.py

```diff
@@ -1897,10 +1905,22 @@ def natural_join(
             None,
         )
         if self._select_statement:
-            select_plan = SelectStatement(
-                from_=SelectSnowflakePlan(join_plan, analyzer=self._session._analyzer),
-                analyzer=self._session._analyzer,
-            )
+            if isinstance(self._select_statement, MockSelectable):
+                select_plan = MockSelectStatement(
+                    projection=[Star([])],
+                    from_=MockSelectExecutionPlan(
+                        join_plan,
+                        analyzer=self._session._analyzer,
+                    ),
+                    analyzer=self._session._analyzer,
+                )
+            else:
+                select_plan = SelectStatement(
+                    from_=SelectSnowflakePlan(
+                        join_plan, analyzer=self._session._analyzer
+                    ),
+                    analyzer=self._session._analyzer,
+                )
```

Comment:
do you think we can move the logic into mock analyzer, into those `create_*` methods?

I was previously discussing with Yijun that we want to introduce as few changes in `dataframe` as possible to keep the mocking related feature modularized/isolated.